protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,233 @@
+# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)
+
+## Context
+
+### Consensus Properties
+
+Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following properties.
+For every height $h$:
+
+- _agreement_: no two correct processes decide differently.
+- _validity_: function $valid(v, s)$, when $v$ is the decided block, always returns _true_.
+- _termination_: all correct processes eventually decide.
+
+The _validity_ property refers to $valid(v, bc_{h-1})$, which is defined as a (mathematical) function
+whose output depends exclusively on its inputs: the value proposed $v$ (i.e., a block),
+and the state of the blockchain after applying the block decided at height $h-1$,
+denoted $bc_{h-1}$.
+
+The consensus algorithm implemented in CometBFT (Tendermint) fulfills these properties.
+
+### ABCI interface
+
+In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
+the implementation of function $valid(v, bc_{h-1})$ was totally internal to CometBFT.
+Technically, the application's part of $bc_{h-1}$ was not considered by function $valid()$.
+Thus, the application had no direct say on the validity of a block,
+although it could (and still can) indirectly influence the contents of blocks via the (best-effort) ABCI call `CheckTx`.
+
+With the evolution of ABCI to ABCI 1.0 and 2.0, CometBFT's implementation of
+function $valid(v, bc_{h-1})$ now has two components:
+
+- the validity checks performed directly by CometBFT on blocks
+  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
+- the validity checks that now the application can perform as part of `ProcessProposal`;
+  i.e., `ProcessProposal` is now part of $valid(v, bc_{h-1})$
+
+With the new structure of the implementation of function $valid(v, bc_{h-1})$:
+
+- consensus _agreement_ is not affected and all processes are still required to agree on the same value
+- the consensus _validity_ property is not affected since we are changing the
+  internals of function $valid(v, bc_{h-1})$; consensus _validity_ just requires this function to be true
+
+However, the new structure of the implementation of function $valid(v, bc_{h-1})$
```

Comment:
Termination can only be achieved (while satisfying the other properties) if valid proposals are made. This has not changed with the new specification. What has changed, as you have pointed elsewhere, is that valid is possibly more restrictive now. Maybe this should be clarified here.

Reply:
Added sentence to clarify